src/VecSim/algorithms/hnsw/hnsw.h

```diff
@@ -1007,24 +1018,35 @@ void HNSWIndex<DataType, DistType>::repairConnectionsForDeletion(
         }
     } else {
         // We don't need to filter the candidates - just update the edges.
-        neighbor_level.setLinks(candidate_ids);
+        size_t new_num_links = 0;
+        for (idType chosen_cand : candidate_ids) {
+            // We will add the selected element as a new neighbor ONLY IF:
+            // 1. The selected element was already a connection before - hence, no update performed.
+            // 2. Both the neighbour which we are updating and the selected element to connect are
+            // not deleted, hence we do not create new connections for marked deleted elements.
+            if (neighbour_orig_neighbours_set[chosen_cand] ||
+                (!isMarkedDeleted(neighbour_id) && !isMarkedDeleted(chosen_cand))) {
+                neighbor_level.setLinkAtPos(new_num_links++, chosen_cand);
+            }
+        }
+        neighbor_level.setNumLinks(new_num_links);
```

Comment:
aren't we locked here when performing this change? Do we have to check the deletion mark on every iteration?
Also the comment above should be fixed